### PR TITLE
Add method to easily mass-create regions

### PR DESCRIPTION
**Pull request by coasterteam**
A good way of pushing devs towards a likely easier way to setup a large amount of zones. Updated the README to include this as well. Follows code type-checking and format.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -67,6 +67,25 @@ local IsVisible = RegionState:IsRegionVisible("Region1") --Bool for if a region
 local InRegion = RegionState:IsInRegion("Region1", Vector3.new()) --Bool for if a point is in a region
 ```
 
+## Quicker Region Setup
```
Comment (TheNexusAvenger) [inline README.md]: I wouldn't have this as the same type of subsection as `RegionState` and `ModelCulling` when it should be part of `RegionState` or under it.

File: README.md [hunk 2]
```diff
@@ -67,6 +67,25 @@ local IsVisible = RegionState:IsRegionVisible("Region1") --Bool for if a region
...
+|  |-- ZonePart : BasePart
+|- Region2 : Instance
+|  |-- ZonePart : BasePart
+```
+Using the instance above as an example,
+```lua
+local RegionCulling = require(game:GetService("ReplicatedStorage"):WaitForChild("RegionCulling"))
+local RegionState = RegionCulling.RegionState
+
+RegionState:InsertRegionsFromInstance(game.ServerStorage.Regions)
```
Comment (TheNexusAvenger) [inline README.md]: Everything runs on the client. Referencing `ServerStorage` will be... problematic. Also should use `GetService` to be consistent with the other examples.

File: src/State/RegionState.lua [hunk 3]
```diff
@@ -95,6 +95,29 @@ function RegionState:AddRegion(RegionName: string, Center: CFrame, Size: Vector3
...
+		--Get the region name from the child name
+		local RegionName = RegionParent.Name
+		for _, RegionPart in RegionParent:GetChildren() do 
+			if RegionPart:IsA("BasePart") then
+				self:AddRegion(RegionName, RegionPart.CFrame, RegionPart.Size)
+			end
+		end
+	end
+	--// TODO
+	--// Perform region checks to maybe auto-connect intersecting regions?
```
Comment (TheNexusAvenger) [inline src/State/RegionState.lua]: I don't really see the need to handle checking for intersections. You may actually want that like.

File: src/State/RegionState.lua [hunk 4]
```diff
@@ -95,6 +95,29 @@ function RegionState:AddRegion(RegionName: string, Center: CFrame, Size: Vector3
...
+--]]
+function RegionState:InsertRegionsFromInstance(Instances: Model | Folder | Instance): ()
+	if #Instances:GetChildren() == 0 then
+		error("Instance \"" .. Instances:GetFullName() .. "\" has no child instances, unable to build Regions from it.")
+	end
+	for _, RegionParent in Instances:GetChildren() do 
+		--Get the region name from the child name
+		local RegionName = RegionParent.Name
+		for _, RegionPart in RegionParent:GetChildren() do 
+			if RegionPart:IsA("BasePart") then
```
Comment (TheNexusAvenger) [inline src/State/RegionState.lua]: Not requesting a change or anything - just mentioning the rest of the code uses a lot of `if not [condition] then continue end`. It makes the code a bit less like a Christmas tree.

**Review by TheNexusAvenger — COMMENTED** [review]
The code of the changes is fine, but I am not sure about the concept. Innovation Inc Thermal Power Plant stores all regions as data with the bounding boxes made from the relevant models. This design requires you to manage your regions along the models, which may be good or bad. You also then need to manage *another* source of truth to manage which regions can see which.

A unit test for this behavior would also be requested.

**Comment by coasterteam** [comment]
> Innovation Inc Thermal Power Plant stores all regions as data with the bounding boxes made from the relevant models. This design requires you to manage your regions along the models, which may be good or bad.

That is a good way of handling setup. Basing loads on camera location along with zone intersections makes it easier to subdivide the regions. This could be converted to automatically add regions based on model bounding boxes instead of requiring you to build your own zones like this would add. Both could co-exist, but there's a cross-over of use cases.

**Comment by TheNexusAvenger** [comment]
I am still not sure how to feel about dealing with 2 (or 3 if you include the models) sources of truth. You will have your region stored as parts and still need to store the connections somewhere else. The README doesn't provide any guidance on this.

**Comment by coasterteam** [comment]
There's probably a better way to handle the For loop in https://github.com/IITPP-Roblox/Region-Culling/pull/1/commits/a299d47ee66187413cd8411400739036939d6761, not completely used for strict types yet. 

Added a new test case (it does pass!), updated types, tried to re-elaborate on README (still may need to come back to it), updated RegionState & BufferedRegionState to include the method. Created a working use case in my own world as well. 

Please let me know if there is still anything to resolve.

File: README.md [hunk 5]
```diff
@@ -39,8 +39,8 @@ RegionState:SetVisibleWhenOutsideRegions("Region3")
 
 --Make it so region 1 can see 2 and 3.
 --ConnectRegions makes it so both regions can see each other.
-RegionStateObject:ConnectRegions("Region1", "Region2")
-RegionStateObject:ConnectRegions("Region1", "Region3")
+RegionState:ConnectRegions("Region1", "Region2")
+RegionState:ConnectRegions("Region1", "Region3")
```
Comment (TheNexusAvenger) [inline README.md:43]: (No change) This was a good find. I completely missed that when copying from the unit test.

File: README.md [hunk 6]
```diff
@@ -67,6 +67,28 @@ local IsVisible = RegionState:IsRegionVisible("Region1") --Bool for if a region
...
+|  |-- ZonePart : BasePart
+```
+Using the instance above as an example,
+```lua
+local ReplicatedStorage = game:GetService("ReplicatedStorage")
+local RegionCulling = require(ReplicatedStorage:WaitForChild("RegionCulling"))
+local RegionState = RegionCulling.RegionState
+
+RegionState:InsertRegionsFromInstance(ReplicatedStorage:WaitForChild("Regions"))
+--// RegionCulling will generate all regions based on the provided instances.
```
Comment (TheNexusAvenger) [inline README.md:89]: Not a fan of a future-tense comment after the action instead of a present-tense comment before. `--//` is also not used in any other code blocks.

File: src/State/RegionState.lua [hunk 7]
```diff
@@ -95,6 +95,28 @@ function RegionState:AddRegion(RegionName: string, Center: CFrame, Size: Vector3
...
+		error("Instance \"" .. Instances:GetFullName() .. "\" has no child instances, unable to build Regions from it.")
+	end
+	for _, RegionParent in Instances:GetChildren() do 
+		--Get the region name from the child name
+		local RegionName = RegionParent.Name
+		for _, Inst in RegionParent:GetChildren() do 
+			if not Inst:IsA("BasePart") then
+				continue
+			end
+			local RegionPart : BasePart = Inst :: BasePart
```
Comment (TheNexusAvenger) [inline src/State/RegionState.lua:114]: A) Is using `: BasePart` on the left side needed when `:: BasePart` is on the right side?
B) If the `: BasePart` is needed, the rest of the code does not have a leading space.

File: src/State/RegionState.lua [hunk 8]
```diff
@@ -95,6 +95,28 @@ function RegionState:AddRegion(RegionName: string, Center: CFrame, Size: Vector3
     })
 end
 
+--[[
+Creates regions based on a pre-determined model/folder, provides for quicker setup &
+easier customization. 
+Name each region with the appropriate title and then add parts to determine their region.
+--]]
+function RegionState:InsertRegionsFromInstance(Instances: Model | Folder | Instance): ()
```
Comment (TheNexusAvenger) [inline src/State/RegionState.lua:103]: Typing in `Types.lua` disagrees with `Model | Folder | Instance`. `Instance` should cover this without any issue.

File: src/State/BufferedRegionState.lua [hunk 9]
```diff
@@ -92,6 +92,15 @@ function BufferedRegionState:AddRegion(RegionName: string, Center: CFrame, Size:
     self.WrappedRegionState:AddRegion(RegionName, Center, Size)
 end
 
+--[[
+Creates regions based on a pre-determined model/folder, provides for quicker setup &
+easier customization. 
+Name each region with the appropriate title and then add parts to determine their region.
+--]]
+function BufferedRegionState:InsertRegionsFromInstance(Instances: Model | Folder | Instance): ()
```
Comment (TheNexusAvenger) [inline src/State/BufferedRegionState.lua:100]: Typing in `Types.lua` disagrees with `Model | Folder | Instance`. `Instance` should cover this without any issue.

File: test/State/RegionState.spec.lua [hunk 10]
```diff
@@ -142,5 +142,19 @@ return function()
...
+            local Region4 = Instance.new("Model", InstanceTree)
+            Region4.Name = "Region4"
+            local BasePart = Instance.new("Part", Region4)
+            BasePart.CFrame = CFrame.new(0, 4, 0)
+            BasePart.Size = Vector3.new(2, 2, 2)
+                    
+            RegionStateObject:InsertRegionsFromInstance(InstanceTree)
+            expect(RegionStateObject:IsInRegion("Region4", Vector3.new(0, 1, 0))).to.equal(false)
+            expect(RegionStateObject:IsInRegion("Region4", Vector3.new(0, 3, 0))).to.equal(true)
+        end)
```
Comment (TheNexusAvenger) [inline test/State/RegionState.spec.lua:158]: Would it be reasonable to add a test case for an empty instance? `expect(function() RegionStateObject:InsertRegionsFromInstance(Instance.new("Folder") end).to.throw("Instance \"Folder\" has no child instances, unable to build Regions from it.")` should work.

**Review by TheNexusAvenger — COMMENTED** [review]
From the previous comment, I am still not sure about this idea. While it does regions, you will still need to manage a second source of truth for managing the connections.
